### PR TITLE
Add extend and split methods to Semaphore guards

### LINKED ISSUE #1: Permit arithmetic
Your crate is very close to what I need.  However, I need one thing more: the ability to add and subtract permits.  For example, I need to be able to acquire a `SemaphoreGuard` with 10 permits and another with 5, then combine them into a single `SemaphoreGuard` with 15.  I also need to be able to do the reverse: split a single `SemaphoreGuard` into two, with a selectable division of permits between them.  Do you think this feature would be a useful addition to your library?  If so, I'll try to add it.

### PR DESCRIPTION
**Pull request by asomers**
extend allows combining two guards into one, with the combined number of
permits.

split allows splitting one guard into two, with a selectable division of
permits.

Fixes #1

### CODE REVIEW

File: src/guard.rs [hunk 1]
```diff
@@ -63,12 +90,63 @@ impl<'a> SemaphoreGuard<'a> {
...
+
+    /// Combine two `SemaphoreGuardArc`s into one, with the sum of the originals' permits.
+    ///
+    /// # Examples
+    /// ```
+    /// # use async_weighted_semaphore::Semaphore;
+    /// # tokio_test::block_on(async {
+    /// let semaphore = Semaphore::new(15);
+    /// let mut g1 = semaphore.acquire_arc(10).await.unwrap();
+    /// let g2 = semaphore.acquire_arc(5).await.unwrap();
```
Comment (nathdobson) [inline src/guard.rs:132]:
```suggestion
    /// let mut g1 = semaphore.acquire(10).await.unwrap();
    /// let g2 = semaphore.acquire(5).await.unwrap();
```

File: src/guard.rs [hunk 2]
```diff
@@ -31,10 +31,37 @@ pub struct SemaphoreGuardArc {
...
+    /// let mut g1 = semaphore.acquire(10).await.unwrap();
+    /// let g2 = semaphore.acquire(5).await.unwrap();
+    /// g1.extend(g2);
+    /// # })
+    /// ```
+    pub fn extend(&mut self, other: SemaphoreGuard<'a>) {
+        if std::ptr::eq(self.semaphore, other.semaphore) {
+            self.amount += other.forget();
+        } else {
+            self.semaphore.poison();
```
Comment (nathdobson) [inline src/guard.rs]: I would expect a panic here - it's a very strange case.

File: src/guard.rs [hunk 3]
```diff
@@ -31,10 +31,37 @@ pub struct SemaphoreGuardArc {
...
+    /// # tokio_test::block_on(async {
+    /// let semaphore = Semaphore::new(15);
+    /// let mut g1 = semaphore.acquire(10).await.unwrap();
+    /// let g2 = semaphore.acquire(5).await.unwrap();
+    /// g1.extend(g2);
+    /// # })
+    /// ```
+    pub fn extend(&mut self, other: SemaphoreGuard<'a>) {
+        if std::ptr::eq(self.semaphore, other.semaphore) {
+            self.amount += other.forget();
```
Comment (nathdobson) [inline src/guard.rs]: Need to handle overflow here.

**Review by nathdobson — COMMENTED** [review]
Sorry for the delay. Mostly looks good.

File: src/guard.rs (hunk 3 above)
Comment (asomers) [reply src/guard.rs]: Actually, it's handled automatically.  I'm not sure where, but somehow the semaphore gets poisoned on overflow.  See `test_extend_overflow`.

Comment (nathdobson) [reply src/guard.rs]: The addition operation itself could overflow if extend is called multiple times 8 times with MAX_AVAILABLE.

**Review by nathdobson — APPROVED** [review]

**Comment by nathdobson** [comment]
This should now be available in version 0.2.1 on crates.io.

**Pull request merged**